articles/execute_process_refactor.md

```diff
@@ -0,0 +1,619 @@
+
+Additional properties provide access to constructor parameters for `process_description`, `shell`, and `output`.
+
+#### Methods
+
+| Name               | Description                                                  |
+| ------------------ | ------------------------------------------------------------ |
+| get_sub_entities   | Override. If on_exit was provided in the constructor, returns that; otherwise returns an empty list. |
+| execute            | Override. Establishes event handlers for process execution, passes the execution context to the process definition for substitution expansion, then uses `osrf_pycommon.process_utils.async_execute_process` to launch the defined process. |
+| get_asyncio_future | Override. Return an asyncio Future, used to let the launch system know when we're done. |
+
+#### Events
+
+##### Handled Events
+
+| Event Type                              | Description                                                  |
+| --------------------------------------- | ------------------------------------------------------------ |
+| `launch.events.process.ShutdownProcess` | Begins standard shutdown procedure for a running executable  |
+| `launch.events.process.SignalProcess`   | Passes the signal provided by the event to the running process |
+| `launch.events.process.ProcessStdin`    | Passes the text provided by the event to the stdin of the process |
+| `launch.events.Shutdown`                | Same as ShutdownProcess                                      |
+
+##### Emitted Events
+
+| Event Type                             | Description                                                  |
+| -------------------------------------- | ------------------------------------------------------------ |
+| `launch.events.process.ProcessStarted` | Emitted when the process starts                              |
+| `launch.events.process.ProcessExited`  | Emitted when the process exits; event contains return code   |
+| `launch.events.process.ProcessStdout`  | Emitted when the process produces data the stdout pipe; event contains the data from the pipe |
+| `launch.events.process.ProcessStderr`  | Emitted when the process produces data the stderr pipe; event contains the data from the pipe |
+
+### launch_ros.descriptions.Node
+
+This class would represent only the common information required to define a ROS2 node.
+It would not inherit from any other class, and would have a subset of the fields of the current `launch_ros.actions.Node`.
+
+#### Constructor
+
+These parameters are drawn from the shared parameters of the current `launch_ros.actions.Node` and `launch_ros.descriptions.ComposableNode`, and introduces a new collection of `traits`, described in following classes.
+
+| Argument       | Description                                                  |
+| -------------- | ------------------------------------------------------------ |
+| package        | name of the ROS package the node executable lives in         |
```

Comment:
@roger-strain hmm, I'd rather say ROS executables and node plugins live in packages, not nodes. So I'm not sure if it makes sense for node descriptions to refer to a package.

Reply:
I've adjusted this now as well. `package` has moved to the `RosExecutable` object, and been re-introduced for `ComposableNode`. I also want the one on `ComposableNode` to be optional, so it can be picked up from the executable context if possible. If that's not reasonable, we can remove that concept. Would be nice to not have to respecify the package for each composed node. As I'm writing that, it feels wrong, so I think it might have to come back out. Let me know what you think.

Reply:
> I also want the one on ComposableNode to be optional, so it can be picked up from the executable context if possible. If that's not reasonable, we can remove that concept. Would be nice to not have to respecify the package for each composed node.

That's interesting sugar. I like it.